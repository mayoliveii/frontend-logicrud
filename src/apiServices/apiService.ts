/* eslint-disable @typescript-eslint/no-explicit-any */
import axios from 'axios';
import env from '../../env';

axios.create({
  baseURL: env.VITE_API_URL,
});

export interface NewCategoryData {
  title: string;
  description: string;
}

export interface EditCategoryData {
  id: string;
  title: string;
  description: string;
}

export interface CategoryData {
  id: string;
  title: string;
  description: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface ListCategoriesParams {
  orderBy?: 'created_at_ASC' | 'created_at_DESC';
  startDate?: string;
  endDate?: string;
  search?: string;
}

const api = axios.create({
  baseURL: env.VITE_API_URL,
});

const apiService = {
  getAllCategories: async (params: ListCategoriesParams): Promise<CategoryData[]> => {
    try {
      const response = await api.get<CategoryData[]>('/list-categories', { params });

      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  createCategory: async (categoryData: NewCategoryData) => {
    try {
      const response = await api.post('/create-category', categoryData);
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  editCategory: async (categoryData: EditCategoryData) => {
    try {
      const response = await api.post(`/edit-category`, categoryData);
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  getCategoryById: async (categoryId: string, categoryData: CategoryData) => {
    try {
      const response = await api.post(`/get-category-by-id?id=${categoryId}`, categoryData);
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  deleteCategory: async (categoryId: string) => {
    try {
      const response = await api.delete(`/delete-category`, { data: { id: categoryId } });
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  searchCategorys: async (searchTerm: string, params: ListCategoriesParams) => {
    try {
      const response = await api.get('/list-categories', { params: { ...params, search: searchTerm } });
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
};

const handleApiError = (error: any) => {
  if (error.response) {
    throw new Error(`Error from server: ${JSON.stringify(error.response.data)}`);
  }

  if (error.request) {
    throw new Error(`No response from server: ${error.request}`);
  }

  if (error.message) {
    throw new Error(`Error: ${error.message}`);
  }
};

export default apiService;
