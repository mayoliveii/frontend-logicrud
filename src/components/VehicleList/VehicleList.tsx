import { useCallback, useEffect, useState } from 'react';
import apiService, { CategoryData } from '../../apiServices/apiService';
import { Alert, AlertIcon, Button, Center, CircularProgress, CloseButton, Container, Flex, useDisclosure } from '@chakra-ui/react';
import Vehicle from '../Vehicle/Vehicle';
import { Input } from "@chakra-ui/react";
import NotFoundMessage from '../NotFoundMessage/NotFoundMessage';
import { CreateModal } from '../CreateModal/CreateModal';
import "react-datepicker/dist/react-datepicker.css";
import DatePicker from "react-datepicker";
import { FaRegCalendarAlt } from 'react-icons/fa';
import React from 'react';

export interface ModalProps {
  onCategoryAdded?: () => void;
  onCategoryDeleted?: () => void;
  onCategoryUpdated?: () => void;
}

export default function VehicleList() {
  const { isOpen, onOpen, onClose } = useDisclosure();

  const [categories, setCategories] = useState<CategoryData[]>([]);
  const [searchTerm, setSearchTerm] = useState("");
  const [startDate, setStartDate] = useState<Date | null>(null);
  const [endDate, setEndDate] = useState<Date | null>(null);
  const [isDateFilterApplied, setDateFilterApplied] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [showAlert, setShowAlert] = useState(false);

  const fetchCategories = useCallback(async () => {
    setIsLoading(true);
    const timeoutId = setTimeout(() => {
      setShowAlert(true);
    }, 5000); // Mostrar alerta apÃ³s 5 segundos

    try {
      const allCategories = await apiService.getAllCategories({
        orderBy: 'created_at_DESC',
        startDate: startDate ? startDate.toISOString() : undefined,
        endDate: endDate ? endDate.toISOString() : undefined
      });
      clearTimeout(timeoutId);
      setCategories(allCategories);
      setDateFilterApplied(!!startDate || !!endDate);
    } catch (error) {
      clearTimeout(timeoutId);
      console.error(error);
    } finally {
      setIsLoading(false);
    }
  }, [startDate, endDate]);

  useEffect(() => {
    fetchCategories();
  }, [fetchCategories]);

  const updateVehicleList = () => {
    fetchCategories();
  }

  const clearFilters = () => {
    setStartDate(null);
    setEndDate(null);
    setSearchTerm("");
    setDateFilterApplied(false);
  };

  const filteredCategories = categories.filter(
    category =>
      category.title.toLowerCase().includes(searchTerm.toLowerCase()) ||
      category.id.toLowerCase().includes(searchTerm.toLowerCase())
  );

  const CustomInput = React.forwardRef<HTMLInputElement, { value?: string, onClick?: () => void, placeholder?: string }>(
    ({ value, onClick, placeholder }, ref) => (
      <div style={{ display: 'flex', alignItems: 'center', position: 'relative' }}>
        <FaRegCalendarAlt style={{ position: 'absolute', left: '10px', top: '15px', color: '#FF7A50' }} />
        <input
          style={{
            padding: '11px',
            paddingLeft: '30px',
            fontSize: '16px',
            borderRadius: '25px',
            border: '1px solid #FF7A50',
            width: '150px'
          }}
          onClick={onClick}
          value={value}
          placeholder={placeholder}
          ref={ref}
        />
      </div>
    )
  );

  return (
    <Container maxW={'5xl'} my={20}>
      {showAlert && isLoading && (
        <Alert status='warning'>
          <AlertIcon />
          Estamos notando um atraso inesperado na resposta do nosso banco de dados.
          Por favor, aguarde mais alguns segundos enquanto carregamos a lista para vocÃª.
          Agradecemos a sua paciÃªncia.
        </Alert>
      )}
      <Flex my={8} align="center" justifyContent="space-between">
        <Button
          size={'lg'}
          fontWeight={'normal'}
          px={6}
          onClick={onOpen}
          colorScheme='orange'
          variant='outline'
        >
          Nova categoria
        </Button>
        <CreateModal isOpen={isOpen} onClose={onClose} onCategoryAdded={updateVehicleList} />

        <div style={{ display: 'flex', alignItems: 'center' }}>
          <div style={{ position: 'relative' }}>
            <DatePicker
              className="myDatePicker"
              placeholderText="Start date"
              selected={startDate}
              onChange={setStartDate}
              selectsStart
              startDate={startDate}
              endDate={endDate}
              onFocus={() => true}
              customInput={<CustomInput />}
            />
          </div>
          <div style={{ position: 'relative', marginLeft: '15px' }}>
            <DatePicker
              className="myDatePicker"
              placeholderText="End date"
              selected={endDate}
              onChange={setEndDate}
              selectsEnd
              startDate={startDate}
              endDate={endDate}
              minDate={startDate}
              onFocus={() => true}
              customInput={<CustomInput />}
            />
          </div>
          <CloseButton size="lg" ml={3} rounded={30} _hover={{ bg: 'orange.100' }}
            onClick={clearFilters} variant={'ghost'} />
        </div>
      </Flex>
      <Input
        placeholder='Pesquisar veÃ­culo por tÃ­tulo ou ID'
        size='md'
        focusBorderColor="#FF7A50"
        value={searchTerm}
        onChange={e => setSearchTerm(e.target.value)}
      />
      {isLoading ? (
        <Center ml={'auto'} h={'50vh'}>
          <CircularProgress isIndeterminate color='orange.500' />
        </Center>
      ) : filteredCategories.length > 0 ? (
        filteredCategories.map(category => (
          <Vehicle key={category.title} category={category} onCategoryDeleted={updateVehicleList} onCategoryUpdated={updateVehicleList} />
        ))
      ) : (isDateFilterApplied || categories.length > 0) ? (
        <NotFoundMessage
          title="Ops! Categoria de veÃ­culo nÃ£o encontrada ðŸ‘€"
          description="A categoria de veÃ­culo que vocÃª tentou acessar nÃ£o estÃ¡ disponÃ­vel em nosso sistema.
    NÃ£o encontramos nenhum registro correspondente ao tÃ­tulo ou ID fornecido.
    Continue explorando nosso sistema para descobrir outras categorias disponÃ­veis."
        />
      ) : (
        <NotFoundMessage
          title="Nenhuma categoria de veÃ­culo cadastrada ðŸ˜°"
          description="NÃ£o hÃ¡ categorias de veÃ­culos disponÃ­veis em nosso sistema no momento. Por favor, adicione uma nova categoria."
        />
      )
      }
    </Container>
  );
}