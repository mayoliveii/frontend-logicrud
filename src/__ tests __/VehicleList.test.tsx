jest.mock('react-datepicker/dist/react-datepicker.css', () => { });

import { render } from '@testing-library/react';
import { ChakraProvider } from '@chakra-ui/react';
import VehicleList from '../components/VehicleList/VehicleList';
import '@testing-library/jest-dom';
import apiService from '../apiServices/apiService';

jest.mock('./../apiServices/apiService', () => ({
  getAllCategories: jest.fn(),
}));

describe('VehicleList', () => {
  it('renders VehicleList component and matches snapshot', () => {
    (apiService.getAllCategories as jest.Mock).mockResolvedValue([]);
    const { asFragment } = render(
      <ChakraProvider>
        <VehicleList />
      </ChakraProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the input search', () => {
    (apiService.getAllCategories as jest.Mock).mockResolvedValue([]);
    const { getByPlaceholderText } = render(
      <ChakraProvider>
        <VehicleList />
      </ChakraProvider>
    );
    expect(getByPlaceholderText('Pesquisar veículo por título ou ID')).toBeInTheDocument();
  });
});