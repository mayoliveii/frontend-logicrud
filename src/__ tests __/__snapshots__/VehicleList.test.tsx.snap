// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleList renders VehicleList component and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-1idzpo0"
  >
    <div
      class="css-g0w66m"
    >
      <button
        class="chakra-button css-yxc4db"
        type="button"
      >
        Nova categoria
      </button>
      <div
        style="display: flex; align-items: center;"
      >
        <div
          style="position: relative;"
        >
          <div
            class="react-datepicker-wrapper"
          >
            <div
              class="react-datepicker__input-container"
            >
              <div
                style="display: flex; align-items: center; position: relative;"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: rgb(255, 122, 80); position: absolute; left: 10px; top: 15px;"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
                  />
                </svg>
                <input
                  placeholder="Start date"
                  style="padding: 11px 11px 11px 30px; font-size: 16px; border-radius: 25px; border: 1px solid #ff7a50; width: 150px;"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          style="position: relative; margin-left: 15px;"
        >
          <div
            class="react-datepicker-wrapper"
          >
            <div
              class="react-datepicker__input-container"
            >
              <div
                style="display: flex; align-items: center; position: relative;"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: rgb(255, 122, 80); position: absolute; left: 10px; top: 15px;"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
                  />
                </svg>
                <input
                  placeholder="End date"
                  style="padding: 11px 11px 11px 30px; font-size: 16px; border-radius: 25px; border: 1px solid #ff7a50; width: 150px;"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <button
          aria-label="Close"
          class="css-w37fj6"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <input
      class="chakra-input css-uofnrp"
      placeholder="Pesquisar veículo por título ou ID"
      value=""
    />
    <div
      class="css-1qq1sqf"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="chakra-progress css-120wkjd"
        data-indeterminate=""
        role="progressbar"
      >
        <svg
          class="css-jf70f3"
          viewBox="0 0 100 100"
        >
          <circle
            class="chakra-progress__track css-ol3i12"
            cx="50"
            cy="50"
            r="42"
            stroke-width="10px"
          />
          <circle
            class="chakra-progress__indicator css-u3myz4"
            cx="50"
            cy="50"
            r="42"
            stroke-width="10px"
          />
        </svg>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
